### ISSUE TITLE
Use Kafka client specific consumer/producer config properties

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#2904]
Currently, the configuration of the Kafka consumer for receiving command messages in the Command Router component is done via the generic Kafka configuration properties:
```
hono:
  kafka:
    consumerConfig:
      bootstrap.servers: [...]
      [...]
```

Especially with the introduction of another kind of Kafka consumer in the Command Router (see #2837), there should be a way to configure the Kafka *command* consumer explicitly.

The Command Router Kafka command consumer config needs:
- standard Kafka config properties passed on to the Kafka consumer
- specific properties introduced by the Hono consumer classes, like `HonoKafkaConsumer#POLL_TIMEOUT` and the `AsyncHandlingAutoCommitKafkaConsumer#CONFIG[*]` properties
(see also the Hono `KafkaConsumerConfigProperties` class, currently not used for these classes)

The definition of the new config structure should be done taking into account how the AMQP messaging network command consumer is configured. This is currently done like this:
```
hono:
  command:
    name: 'Hono Command Router'
    host: someHost
    tlsEnabled: true
    trustStorePath: /etc/hono/trusted-certs.pem
    initialCredits: 2000
    sendMessageTimeout: 5000
    linkEstablishmentTimeout: 1000
```

**Comment by b-abel** [comment]
What about using the same keywords as with AMQP with the difference that they are not prefixed by `hono.` but instead by `hono.kafka.`? For example: `hono.kafka.messaging` and  `hono.kafka.command`. 

~~~yaml
hono:
  kafka:
    command: 
      pollTimeout: 50L
      commonClientConfig:
        bootstrap.servers: ${hono.kafka.bootstrap.servers}
~~~

**Comment by b-abel** [comment]
This could be considered breaking change because the components will no longer work with an existing configuration which also affects northbound clients like the CLI. On the other hand, I think there is no need to wait for Hono 2.0 because it only affects the Kafka support which is still under development. For me, the only consequence of this is that we should not rush this and avoid that we have to make changes in the structure of the configuration again in one of the next versions. 
WDYT @sophokles73, @calohmn?

**Comment by b-abel** [comment]
My proposal above could for example look like the following (MQTT adapter, configured for AMQP _and_ Kafka):

~~~yaml
hono:
  messaging:
    name: Hono adapter-mqtt-vertx
    amqpHostname: hono-internal
    host: eclipse-hono-dispatch-router
    port: 5673
    keyPath: /etc/sandbox/hono.eclipseprojects.io-key.pem
    certPath: /etc/sandbox/hono.eclipseprojects.io-cert.pem
    trustStorePath: /etc/hono/trusted-certs.pem
    hostnameVerificationRequired: false
  command:
    name: Hono adapter-mqtt-vertx
    amqpHostname: hono-internal
    host: eclipse-hono-dispatch-router
    port: 5673
    keyPath: /etc/sandbox/hono.eclipseprojects.io-key.pem
    certPath: /etc/sandbox/hono.eclipseprojects.io-cert.pem
    trustStorePath: /etc/hono/trusted-certs.pem
    hostnameVerificationRequired: false
  kafka:
    messaging:
      defaultClientIdPrefix: adapter-mqtt-vertx
      producerConfig:
        bootstrap.servers: eclipse-hono-kafka-0.eclipse-hono-kafka-headless.hono:9094
        security.protocol: SASL_SSL
        sasl.mechanism: SCRAM-SHA-512
        sasl.jaas.config: "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"hono\" password=\"hono-secret\";"
        ssl.truststore.location: /etc/hono/truststore.jks
        ssl.truststore.password: honotrust
        ssl.endpoint.identification.algorithm: ""
    command:
      defaultClientIdPrefix: adapter-mqtt-vertx
      # used by the command consumer and the Kafka Admin client that creates internal command topics
      commonClientConfig:
        bootstrap.servers: eclipse-hono-kafka-0.eclipse-hono-kafka-headless.hono:9094
        security.protocol: SASL_SSL
        sasl.mechanism: SCRAM-SHA-512
        sasl.jaas.config: "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"hono\" password=\"hono-secret\";"
        ssl.truststore.location: /etc/hono/truststore.jks
        ssl.truststore.password: honotrust
        ssl.endpoint.identification.algorithm: ""
~~~

**Comment by b-abel** [comment]
On second thought, I don't think it's such a good idea to use the same configuration structure as for AMQP. The central goal is that each Kafka client can be configured separately. This is important because Kafka clients can behave very differently depending on the configuration and it is to be expected that timeouts etc. have to be changed when operating. For C&C, for example, there are 3 clients (consumer, command sender, and admin client) on the protocol adapter that should not share all configuration properties.
A second, subordinate, goal could be that the connection parameters do not have to be completely repeated, i.e. not like with AMQP where you have to specify host, port, certificates, etc. for (downstream) messaging and C&C respectively. 
@sophokles73 @calohmn WDYT?

**Comment by sophokles73** [comment]
> For C&C, for example, there are 3 clients (consumer, command sender, and admin client) on the protocol adapter that should not share all configuration properties.

Isn't that possible with your latest proposal?
```yaml
kafka:
  command:
    defaultClientIdPrefix: adapter-mqtt-vertx
    # used by the command consumer and the Kafka Admin client that creates internal command topics
    commonClientConfig:
      bootstrap.servers: eclipse-hono-kafka-0.eclipse-hono-kafka-headless.hono:9094
      security.protocol: SASL_SSL
      sasl.mechanism: SCRAM-SHA-512
      sasl.jaas.config: "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"hono\" password=\"hono-secret\";"
      ssl.truststore.location: /etc/hono/truststore.jks
      ssl.truststore.password: honotrust
      ssl.endpoint.identification.algorithm: ""
    producerConfig:
      propery: value
    consumerConfig:
      propery: value
    adminClientConfig:
      propery: value
```

**Comment by b-abel** [comment]
True. But what if command handling would involve two consumers one day? Or, more realistic, separate producers should be used for _event_ and _telemetry_? I would use a separate name for each Kafka client.

**Comment by sophokles73** [comment]
Can you make a concrete proposal then?

**Comment by b-abel** [comment]
**Proposal'**

To configure clients individually, they need separate "names", i.e. prefixes. This could be e.g. at a protocol adapter: `command`, `commandResponse`, `commandInternal`. For telemetry and event, I would use `messaging`. 
I would pull the common configuration to the top level so that it can be used for all clients. That means that it is no longer included in the class `AbstractKafkaConfigProperties`, but a separate implementation, which is then set on the other ConfigurationProperties.

Example protocol adapter config:

~~~yml
hono:
  kafka:
    commonClientConfig:
      bootstrap.servers: eclipse-hono-kafka-0.eclipse-hono-kafka-headless.hono:9094
      security.protocol: SASL_SSL
      sasl.mechanism: SCRAM-SHA-512
      sasl.jaas.config: "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"hono\" password=\"hono-secret\";"
      ssl.truststore.location: /etc/hono/truststore.jks
      ssl.truststore.password: honotrust
      ssl.endpoint.identification.algorithm: ""
    messaging:
      producerConfig:
        property: value
    command:
      consumerConfig:
        property: value
    commandResponse:
      producerConfig:
        property: value
    commandInternal:
      adminClientConfig:
        property: value
    registryNotification:
      consumerConfig:
        property: value
~~~

With Spring this means that we need to declare a bean for each prefix (`hono.kafka.commonClientConfig`, `hono.kafka.messaging`, `hono.kafka.command`, `hono.kafka.commandResponse`, `hono.kafka.commandInternal`).
The configuration would be straightforward because having only `hono.kafka.commonClientConfig` defined would be sufficient. All other parts must only be added if a client needs extra tuning. 

I have not worked with the SmallRye ConfigOptions yet. @sophokles73 Would this work well with them?

**Comment by sophokles73** [comment]
What would be the type of the bean that we use the `hono.kafka.commonClientConfig` prefix for? It seems to be different from the type of the other beans, or am I mistaken?

**Comment by b-abel** [comment]
Yes. I would move the property `commonClientConfig` from the current class `AbstractKafkaConfigProperties` into a new one. It could be called for example `KafkaClientConnectionConfigProperties`. 
On the other Kafka config classes, I would then add a setter for this new type.

**Comment by b-abel** [comment]
I am working on this issue and will provide a PR after #2962 is merged.

**Comment by b-abel** [comment]
Resolved by #2972 and #2989.
